### PR TITLE
Block tasks

### LINKED ISSUE #1140: as a senior editor I want to mark a task as 'blocked' until another task has been done
e.g. on https://edit.laws.africa/works/akn/za/act/1957/33/, 
the first amendment task is blocked by the import task,
the second amendment task is blocked by the first,
etc.

### Suggestion
- Add a 'blocked by' field on a task, that takes one or more task numbers as its input
- Disable such a task from being picked up and automatically add a `blocked` label
- Once the blocking tasks have been done, make the task pick up able and remove the label

### PR DESCRIPTION
**Pull request by goose-life**
closes #1140 

TODO:
- [ ] add date to task edit form (CANCELLED)
- [x] disable Unblock in edit form while a task is blocked_by other tasks
- [x] move Blocked tasks to first column
- [ ] filter potential blocking tasks to tasks on this work and related works (currently all in place) (CANCELLED)
- [ ] do magic during bulk import (DO SEPARATELY)
- [x] update blocked tasks when a blocking task is closed
- [ ] enable bulk update of task states

A blocked task:
image
image
image

A task blocked by another task:
image
image
Deselect blocking task from Blocked by list to unblock:
image

A blocking task:
image

Blocked tasks in lists:
image
image
image

Blocked task stats in work list:
image

### CODE REVIEW

File: indigo_api/models/tasks.py
Comment (longhotsummer) [inline indigo_api/models/tasks.py:296]: Just noting that this doesn't allow blocking a task that is pending review.

File: indigo_app/templates/indigo_api/_task_detail.html
Comment (longhotsummer) [inline indigo_app/templates/indigo_api/_task_detail.html:12]:
```suggestion
                <ul class="mt-2">
```

The alert has inner padding, so rather give some space between the intro sentence and the list.

File: indigo_app/templates/indigo_api/_task_detail.html
Comment (longhotsummer) [inline indigo_app/templates/indigo_api/_task_detail.html:262]: You're loading blocked_by twice, rather do it once (saves a DB trip)

File: indigo_app/urls.py
Comment (longhotsummer) [inline indigo_app/urls.py:39]: Perhaps `.../blocking-tasks`? I prefer dashes over underscores in URLs.

File: indigo_app/views/tasks.py
Comment (longhotsummer) [inline indigo_app/views/tasks.py:139]: I think safest to be any task in this place.

File: indigo_app/views/tasks.py
Comment (longhotsummer) [inline indigo_app/views/tasks.py:427]: Also filter by the current country and locality.

File: indigo_app/views/tasks.py
Comment (longhotsummer) [inline indigo_app/views/tasks.py:432]: We're only checking the perm *after* the task has been blocked. Instead, check the perm first and then pass the user and tasks to block on into task.block(), like we do with task.close(), then that method does the business logic.

File: indigo_app/urls.py
Comment (goose-life) [reply indigo_app/urls.py:39]: I used a verb to follow the other patterns;  what do you think of `change-blocking-tasks`? that would also correspond with the view name

File: indigo_api/models/tasks.py
Comment (longhotsummer) [inline indigo_api/models/tasks.py:298]: Use `self.assign_to(None)`?

Comment (goose-life) [reply indigo_api/models/tasks.py:298]: Nice

File: indigo_app/urls.py
Comment (longhotsummer) [reply indigo_app/urls.py:39]: It's a mixed bag. If we're doing REST (which we vaguely are, eg. projects), then the endpoints should actually be nouns and the HTTP verb determines the action. This doesn't work well for assign, unassign, etc. But it does kinda work for blocking tasks -- you're POSTing to the blocking-tasks collection.

If you're strictly using RPC-style naming, like a method name, then it would be a verb as you say.

File: indigo_app/views/tasks.py
Comment (goose-life) [reply indigo_app/views/tasks.py:427]: Is this a performance thing? because tasks have unique ids across places, right? Happy to do it of course, just curious

File: indigo_app/views/tasks.py
Comment (goose-life) [reply indigo_app/views/tasks.py:432]: If I'm understanding you right, you're suggesting that the block transition method on task set the blocked_by tasks on our task in the process of blocking it (and unblock would clear them). 

We also need to make provision for the blocked_by list just being updated without there being a change in state; that can only happen in the view unless I'm mistaken.

So `task.blocked_by.set(blocked_by)` would only be called in the case of there being blocked_by tasks but the transition permission being False (because you can't move from blocked into blocked).

File: indigo_app/views/tasks.py
Comment (longhotsummer) [reply indigo_app/views/tasks.py:427]: It's about being defensive about what you accept and validating the inputs. The user should only be able to specify task ids of tasks in the same place as the current task.

File: indigo_app/views/tasks.py
Comment (longhotsummer) [reply indigo_app/views/tasks.py:432]: Hmm, good point - maybe get around that by allowing blocked -> blocked as a state transition? Otherwise, set the blocked tasks outside of task.blocked(). Otherwise it's duplicate code.

File: indigo_api/models/tasks.py
Comment (longhotsummer) [inline indigo_api/models/tasks.py:300]: Perhaps make this an explicit parameter with a default of None -- python will do the hard work for you.

File: indigo_app/templates/indigo_api/task_form.html
Comment (longhotsummer) [inline indigo_app/templates/indigo_api/task_form.html:130]: Rather than duplicating the form, just wrap them all in the same form, since the `formaction` attribute does the necessary.

File: indigo_app/views/tasks.py
Comment (goose-life) [reply indigo_app/views/tasks.py:427]: I don't love a blocked-to-blocked transition, and we'd have to then special-case the actions on that so that we don't end up with multiple 'user blocked task' actions in a row if they're just updating the list, so I've changed it to set the blocking tasks outside of the block and unblock methods

Comment (longhotsummer) [reply indigo_app/views/tasks.py:427]: Just sanity check the if tree... what if the task is blocked by the user doesn't have necessary perms to make this change?

File: indigo_app/templates/indigo_api/task_form.html
Comment (longhotsummer) [inline indigo_app/templates/indigo_api/task_form.html:123]: This lands up being `if A or B or C and D` == `if A or B or (C and D)` because AND his higher precedence than OR. (AND is like multiply, OR is like add).

As far as I know, you can't use brackets in expressions in django templates. So you may need to move this into the view.

File: indigo_app/views/workflows.py
Comment (longhotsummer) [inline indigo_app/views/workflows.py:60]: There's an opportunity here (and maybe elsewhere) to do something like `... list(Task.STATES) + ['pending_review']` so that there's less typing of magic strings.

File: indigo_app/views/tasks.py
Comment (goose-life) [reply indigo_app/views/tasks.py:429]: Ah. I thought the permission_required as well as disabling the button if the user doesn't have permissions would cover us, but I see the country permission isn't checked (or it's not obvious if it is, maybe I've missed it now).

So because indigo_api.change_task is listed under permission_required, is the only other thing I need to check for the country permission? or should I explicitly check for indigo_api.change_task at the point where I'm setting the blocked_by list?

I'm not super clear on how permission_required works, but I see there's a has_permission on PlaceViewBase which checks for country permission. So should I just say `elif blocked_by and self.has_permission()`, and either raise a permission denied or go ahead and set the blocked_by list at that point?

Once I'm clear on this I can figure out how not to repeat the `set` line three times :grimacing:

Comment (longhotsummer) [reply indigo_app/views/tasks.py:444]: `permission_required` will be checked right at the start of the request, but assume that you don't know how `has_transition_perm` works, so you can't assume that because we've got this far the user can make this change (since has_transition_perm checks both validity and authorization, right?).

Change the `if` tree to be like this, which I think will make it clearer:

```
if blocked_by:
  # user is trying to block this task, is that transition allowed?
  if has_transition_perm(...):
    # can proceed
    ...
  else:
    raise permission denied
else:
  # user is trying to unblock, can that transition proceed?
  if has_transition_perm(..):
    # cool
    ...
  else:
    raise permission denied
```

File: indigo_app/views/tasks.py
Comment (longhotsummer) [inline indigo_app/views/tasks.py:349]: This is a good candidate to move onto the task object, since it doesn't really have anything to do with the view. It could either be done based on the signals that django-fsm fires when transitions happen, or the `cancel` and `close` methods on the task could call it explicitly (which is probably the best).

File: indigo_app/views/tasks.py
Comment (longhotsummer) [inline indigo_app/views/tasks.py:354]: use `.exist()` rather than loading all the data from the db and not using it.

should `self` at the end be `user`?

File: indigo_app/views/tasks.py
Comment (longhotsummer) [reply indigo_app/views/tasks.py:349]: This improves the abstraction: the view is responsible for translating the user's browser request into some action: eg. blocking or unblocking, etc. The task object itself is responsible for managing the impact of that action, not the view. The view should focus on the request, validating permissions, validating input and parameters, and then delegate to the object (or something else) to do the actual business logic. It makes testing and code re-use much easier.

File: indigo_app/views/tasks.py
Comment (goose-life) [reply indigo_app/views/tasks.py]: I was thinking of refactoring the may_{verb} code, but ya they all use the view, not the user

File: indigo_app/views/tasks.py
Comment (goose-life) [reply indigo_app/views/tasks.py]: I was blocked by only having access to the view, not the user, but since refactoring that code this is now possible. I did a little hoop jumping to get the messages to also say which tasks have been updated and unblocked by closing a task, le tmeknow what you think

File: indigo_api/models/tasks.py
Comment (goose-life) [reply indigo_api/models/tasks.py]: this has now been changed

File: indigo_app/views/tasks.py
Comment (goose-life) [reply indigo_app/views/tasks.py]: I've updated this to exclude closed tasks; we might need to filter it further in future so that not too many tasks are loaded.

**Comment by goose-life** [comment]
image

image

File: indigo_api/models/tasks.py
Comment (longhotsummer) [inline indigo_api/models/tasks.py:192]: passing in the user and not the view is much nicer!

**Review by longhotsummer — APPROVED** [review]

**Pull request merged**
